server/app/services/rpc/node_service_pod_handler.rb

```diff
@@ -14,10 +14,21 @@ def cached_pod(service_instance)
         deploy_rev: service_instance.deploy_rev,
         desired_state: service_instance.desired_state
       }
-      @lru_cache.getset(cache_key) {
+      hooks = service_instance.grid_service.hooks
+      pod_hash = @lru_cache.getset(cache_key) {
         debug "pod_cache miss"
         ServicePodSerializer.new(service_instance).to_hash if service_instance.grid_service
       }
+      # remove cached oneshot hooks
+      hooks.dup.each do |hook|
+        if hook.done_for?(service_instance.instance_number)
+          pod_hash[:hooks].delete_if { |h|
```

Comment:
Mutating a cached/serialized blob is a bit of a hack... this is a cache, so IMO the sane way to implement this is as "only cache the serialized service pod if it does not contain oneshot hooks":

```ruby
      return nil unless service_instance.grid_service

      hooks = service_instance.grid_service.hooks

      unless pod_hash = @lru_cache[cache_key]
        debug "pod_cache miss"
        pod_hash = ServicePodSerializer.new(service_instance).to_hash
      end

      has_oneshot_hooks = hooks.any? { |hook|
        hook.oneshot && pod_hash[:hooks].any? { |h|
           h[:type] == hook.type && h[:cmd] == hook.cmd
        }
      }

      if has_oneshot_hooks
        debug "do not cache service #{pod_hash[:service_name]} pod with oneshot hooks"
      else
        @lru_cache[cache_key] = pod_hash
      end

      pod_hash
```

```
172.17.0.1 - - [19/Sep/2017:14:13:43 +0000] "POST /v1/grids/development/stacks HTTP/1.1" 201 1623 0.1662
D, [2017-09-19T14:13:43.531175 #20] DEBUG -- GridServices::Update: not updating service development/hooks-oneshot/redis without changes
172.17.0.1 - - [19/Sep/2017:14:13:43 +0000] "POST /v1/stacks/development/hooks-oneshot/deploy HTTP/1.1" 200 151 0.1393
I, [2017-09-19T14:13:43.582886 #20]  INFO -- StackDeployWorker: removing following services: 
I, [2017-09-19T14:13:43.593367 #20]  INFO -- StackDeployWorker: deploying service development/hooks-oneshot/redis...
I, [2017-09-19T14:13:43.737389 #8]  INFO -- GridServiceSchedulerWorker: starting development/hooks-oneshot/redis deploy
I, [2017-09-19T14:13:43.743659 #8]  INFO -- GridServiceDeployer: starting to deploy development/hooks-oneshot/redis
I, [2017-09-19T14:13:43.826322 #8]  INFO -- GridServiceDeployer: throttling service instance development/hooks-oneshot/redis deploy because of min_health limit (1 instances in-progress)
I, [2017-09-19T14:13:43.831448 #8]  INFO -- GridServiceInstanceDeployer: Deploying service instance development/hooks-oneshot/redis-1 to node core-01 at 2017-09-19 14:13:43 UTC...
D, [2017-09-19T14:13:43.856740 #20] DEBUG -- Rpc::NodeServicePodHandler: pod_cache miss
D, [2017-09-19T14:13:43.869170 #20] DEBUG -- Rpc::ServicePodSerializer: fire development/hooks-oneshot/redis-1 oneshot hook post_start/test-oneshot
D, [2017-09-19T14:13:43.872349 #20] DEBUG -- Rpc::NodeServicePodHandler: do not cache service hooks-oneshot.redis pod with oneshot hooks
```

Reply:
BTW the `h[:type] == hook.type && h[:cmd] == hook.cmd` comparison isn't exactly unique... the hooks should really be serialized to include the `oneshot: true/false` flag.

Reply:
The reason why I chose the mutate cached object path is perf.. without it we don't get any benefits from caching if there are oneshot hooks.